crates/router/src/core/locking.rs

```diff
@@ -0,0 +1,170 @@
+use actix_web::rt::time as actix_time;
+use error_stack::{IntoReport, ResultExt};
+use redis_interface as redis;
+use router_env::logger;
+
+use super::errors::{self, RouterResult};
+use crate::routes;
+
+// pr: should there be 2 enums AcquireLockStatus and ReleaseLockStatus ?
+pub enum LockStatus {
+    Acquired,
+    AlreadyLocked,
+    NotEnabled,
+    AcquireFailedRetriesExhausted,
+    Released,
+    ReleaseFailedRetriesExhausted,
+}
+
+pub enum ActionOnWaitTimeout {}
+
+pub struct LockingInput {
+    pub unique_locking_key: String,
+    pub api_identifier: String,
+    pub action_on_wait_timeout: ActionOnWaitTimeout,
+    pub merchant_id: String,
+}
+
+pub async fn get_key_and_lock_resource(
```

Comment:
Change the function name, `get_key` operation is not happening inside the function.

And if this function only going to works on a key type, can't we have a unit struct for each type.

And `Trait` `LockResource` having a `lock()` method